### ISSUE TITLE
[Q] compatibility with serial driver

### ISSUE THREAD

**Issue by eadmaster** [mariosgit/VFD_MN12832L#1]
I've found this version of the vfd with a serial driver included:
https://it.aliexpress.com/item/1005003334862374.html

Can it be used with this libray or does it need a rewrite?

**Comment by mariosgit** [comment]
The link does not go anywhere usefull.. so I do some guessing. (maybe append a picture)

This Library is for the plain VFD.

There are breakout boards like this
![Screenshot from 2024-06-08 15-17-54](https://github.com/mariosgit/VFD_MN12832L/assets/25881574/f3332e43-78be-4c55-a1ca-384468989e3e)
This should work. It passes the data lines through and provides the high voltages 👍.

If there is a board providing a serial port, then it has, most likely, it's own controller and you need to get info about the protocol it runs. (However if you can erase the firmware on such a controller you can try to use this library on it. :)

**Comment by eadmaster** [comment]
yes, that's exactly the board, maybe the auction is blocked in your country.

Can you help me figuring out the correct wiring?

The pins on the left are:

 - SIN -> MOSI_PIN
 - LAT -> pinLAT
 - SO1 -> left unconnected?
 - HV = High Voltage (should be provided by the board from 5V?)
 - GND
 - CLK -> SCK_PIN 
 - BLK -> pinBLK 
 - SO2  -> left unconnected?
 - EF -> left unconnected?
 - 5V

If it can be of any help, i have some datasheets here, unfortunately in Chinese:
[MN12832L.pdf](https://github.com/user-attachments/files/15749536/MN12832L.pdf)
[MN12832L_pr1-c8(1).pdf](https://github.com/user-attachments/files/15749537/MN12832L_pr1-c8.1.pdf)

**Comment by mariosgit** [comment]
Yes looks good.
I don't know what HV and EF are. Could be enable lines for filament EF? If so it could blank/powersave

Just power it up..
* expect 5W or 1A power draw
* measure the filament the outer pins on the vfd, should have 3V AC.
* measure the high voltage @ vfd - VDD2 right most on the inner block of pins

If these are there it should work...

The vfd has 2 SIN they seem to be bridged, so no greyscale.

**Comment by eadmaster** [comment]
> I don't know what HV and EF are. Could be enable lines for filament EF? If so it could blank/powersave

they also sent me this piece of code, which was meant for the stm32, where these HV and EF pins seems to be used as "PF0_Set" and "PF1_Set", so i guess they cannot be left unconnected.

[main.c.txt](https://github.com/user-attachments/files/15751899/main.c.txt)

**Comment by mariosgit** [comment]
Ohh the secret Chinese code :)

PF0 = EF is just Set at the beginning of their main. So that should turn on something voltage related. Can be connected HIGH or to my pinPWM.
PF1 = HV? is probably connected to the GCP pin @ vfd ? Can you check this ? This would be the pinGCP in my code. That does some pulsing.. I did not quite get what this does I think that is a sort of PWM cycle related to the gray values.

**Comment by eadmaster** [comment]
> PF1 = HV? is probably connected to the GCP pin @ vfd ? Can you check this ? This would be the pinGCP in my code.

not anytime soon, i still considering if getting this module or a [Text-only model like this](https://vi.aliexpress.com/item/1005005578712631.html) (they seem easier to connect and use via serial commands)

**Comment by slabua** [comment]
Following. I have the same display from Aliexpress. But I have no Teensy. Eventually I want to rewrite it in micropython for the Pico but I also have some esp32 laying around to possibly test with arduino.

**Comment by mariosgit** [comment]
For ESP32 you need to replace the IntervalTimer, as it is used in the example, with something else. It calls the MN12832L::refresh function about 1000 times per sec.

ESP32 timer API: https://docs.espressif.com/projects/arduino-esp32/en/latest/api/timer.html#arduino-esp32-timer-api

---

I don't have a **Pico**, I would suggest making a C library for this, due to a lot of bit level manipulation in the drawPixel and refresh functions.
